File: strata-rs/src/lexer.rs
use crate::error::{ParseError, ParseErrorKind, Span};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TokenKind {
    // Literals:
    Null,
    True,
    False,
    Int(i64),
    String(String),
    Bytes(Vec<u8>),

    // Identifiers:
    Ident(String),

    // Punctuation:
    LBrace,
    RBrace,
    LBracket,
    RBracket,
    Colon,
    Comma,

    EOF,
}

pub struct Lexer<'a> {
    input: &'a [u8],
    offset: usize,
    line: usize,
    column: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            input: input.as_bytes(),
            offset: 0,
            line: 1,
            column: 1,
        }
    }

    fn span(&self) -> Span {
        Span {
            offset: self.offset,
            line: self.line,
            column: self.column,
        }
    }

    fn error(&self, kind: ParseErrorKind) -> ParseError {
        ParseError {
            kind,
            span: self.span(),
        }
    }

    fn peek(&self) -> Option<u8> {
        self.input.get(self.offset).copied()
    }

    fn bump(&mut self) -> Option<u8> {
        let byte = self.peek()?;
        self.offset += 1;

        if byte == b'\n' {
            self.line += 1;
            self.column = 1;
        } else {
            self.column += 1;
        }

        Some(byte)
    }

    fn hex_digit(byte: u8) -> Option<u8> {
        match byte {
            b'0'..=b'9' => Some(byte - b'0'),
            b'a'..=b'f' => Some(byte - b'a' + 10),
            b'A'..=b'F' => Some(byte - b'A' + 10),
            _ => None,
        }
    }

    fn skip_ignored(&mut self) {
        loop {
            //skip whitespace
            while matches!(self.peek(), Some(b' ' | b'\n' | b'\r' | b'\t')) {
                self.bump();
            }

            // line comment with //
            if self.peek() == Some(b'/') && self.input.get(self.offset + 1) == Some(&b'/') {
                self.bump();
                self.bump();
                while let Some(current_byte) = self.peek() {
                    self.bump();
                    if current_byte == b'\n' {
                        break;
                    }
                }
                continue;
            }

            // line comment with #
            if self.peek() == Some(b'#') {
                self.bump();
                while let Some(current_byte) = self.peek() {
                    self.bump();
                    if current_byte == b'\n' {
                        break;
                    }
                }
                continue;
            }

            break;
        }
    }

    fn lex_identifier(&mut self) -> TokenKind {
        let start = self.offset;
        self.bump();

        while matches!(
            self.peek(),
            Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_')
        ) {
            self.bump();
        }

        let ident = std::str::from_utf8(&self.input[start..self.offset])
            .unwrap()
            .to_string();

        match ident.as_str() {
            "null" => TokenKind::Null,
            "true" => TokenKind::True,
            "false" => TokenKind::False,
            _ => TokenKind::Ident(ident),
        }
    }

    fn lex_int(&mut self) -> Result<TokenKind, ParseError> {
        let start = self.offset;

        // opt leading '-'
        if self.peek() == Some(b'-') {
            self.bump();
        }

        let mut saw_digit = false;

        while matches!(self.peek(), Some(b'0'..=b'9')) {
            saw_digit = true;
            self.bump();
        }

        // must have at least 1 digit
        if !saw_digit {
            return Err(self.error(ParseErrorKind::IntegerOutOfRange));
        }

        let slice = &self.input[start..self.offset];
        let text = std::str::from_utf8(slice)
            .map_err(|_| self.error(ParseErrorKind::IntegerOutOfRange))?;

        let parsed_value = text
            .parse::<i64>()
            .map_err(|_| self.error(ParseErrorKind::IntegerOutOfRange))?;

        Ok(TokenKind::Int(parsed_value))
    }

    fn lex_bytes(&mut self) -> Result<TokenKind, ParseError> {
        // must start with 0x
        if self.peek() != Some(b'0') || self.input.get(self.offset + 1) != Some(&b'x') {
            return Err(self.error(ParseErrorKind::MalformedBytesLiteral));
        }

        // consume 0x
        self.bump();
        self.bump();

        let hex_start = self.offset;

        while matches!(self.peek(), Some(b'0'..=b'9' | b'a'..=b'f' | b'A'..=b'F')) {
            self.bump();
        }

        let hex_len = self.offset - hex_start;

        // must have even number of hex digits and at least one byte
        if hex_len == 0 || hex_len % 2 != 0 {
            return Err(self.error(ParseErrorKind::MalformedBytesLiteral));
        }

        let hex = &self.input[hex_start..self.offset];
        let mut bytes = Vec::with_capacity(hex_len / 2);

        for i in (0..hex_len).step_by(2) {
            let high_nibble = Self::hex_digit(hex[i])
                .ok_or_else(|| self.error(ParseErrorKind::MalformedBytesLiteral))?;

            let low_nibble = Self::hex_digit(hex[i + 1])
                .ok_or_else(|| self.error(ParseErrorKind::MalformedBytesLiteral))?;

            bytes.push((high_nibble << 4) | low_nibble);
        }

        Ok(TokenKind::Bytes(bytes))
    }

    fn lex_string(&mut self) -> Result<TokenKind, ParseError> {
        self.bump(); // opening '"'

        let mut out = String::new();

        while let Some(current_byte) = self.peek() {
            match current_byte {
                b'"' => {
                    // closing quote
                    self.bump();
                    return Ok(TokenKind::String(out));
                }

                b'\\' => {
                    //escape seq
                    self.bump();
                    let escape_char = self
                        .bump()
                        .ok_or_else(|| self.error(ParseErrorKind::MalformedBytesLiteral))?;

                    match escape_char {
                        b'"' => out.push('"'),
                        b'\\' => out.push('\\'),
                        b'n' => out.push('\n'),
                        b'r' => out.push('\r'),
                        b't' => out.push('\t'),

                        b'u' => {
                            // \uXXXX
                            let mut codepoint = 0u32;

                            for _ in 0..4 {
                                let hex_byte = self.bump().ok_or_else(|| {
                                    self.error(ParseErrorKind::MalformedBytesLiteral)
                                })?;
                                let digit = Self::hex_digit(hex_byte).ok_or_else(|| {
                                    self.error(ParseErrorKind::MalformedBytesLiteral)
                                })?;
                                codepoint = (codepoint << 4) | (digit as u32);
                            }

                            let unicode_char = char::from_u32(codepoint)
                                .ok_or_else(|| self.error(ParseErrorKind::MalformedBytesLiteral))?;
                            out.push(unicode_char);
                        }
                        _ => return Err(self.error(ParseErrorKind::MalformedBytesLiteral)),
                    }
                }

                b'\n' | b'\r' => {
                    // strings cant span lines
                    return Err(self.error(ParseErrorKind::MalformedBytesLiteral));
                }

                _ => {
                    if current_byte >= 0x80 {
                        return Err(self.error(ParseErrorKind::MalformedBytesLiteral));
                    }

                    out.push(current_byte as char);
                    self.bump();
                }
            }
        }

        Err(self.error(ParseErrorKind::MalformedBytesLiteral))
    }

    pub fn next_token(&mut self) -> Result<Token, ParseError> {
        self.skip_ignored();

        let start = self.span();

        let current_byte = match self.peek() {
            Some(b) => b,
            None => {
                return Ok(Token {
                    kind: TokenKind::EOF,
                    span: start,
                });
            }
        };

        let kind = match current_byte {
            b'{' => {
                self.bump();
                TokenKind::LBrace
            }
            b'}' => {
                self.bump();
                TokenKind::RBrace
            }
            b'[' => {
                self.bump();
                TokenKind::LBracket
            }
            b']' => {
                self.bump();
                TokenKind::RBracket
            }
            b':' => {
                self.bump();
                TokenKind::Colon
            }
            b',' => {
                self.bump();
                TokenKind::Comma
            }

            // string literal
            b'"' => self.lex_string()?,

            // bytes literal
            b'0' if self.input.get(self.offset + 1) == Some(&b'x') => self.lex_bytes()?,

            // integer literal
            b'-' | b'0'..=b'9' => self.lex_int()?,

            // identifier or keyword
            b'a'..=b'z' | b'A'..=b'Z' | b'_' => self.lex_identifier(),

            _ => {
                return Err(self.error(ParseErrorKind::UnexpectedToken {
                    expected: "valid token",
                    found: "invalid character",
                }));
            }
        };

        Ok(Token { kind, span: start })
    }
}

File: strata-js/src/lexer.ts
import { ParseError } from "./parse_error.js";

export type Span = {
    offset: number;
    line: number;
    column: number;
};

export type Token = {
    kind: TokenKind;
    span: Span;
};

export type TokenKind =
    | { type: "Null" }
    | { type: "True" }
    | { type: "False" }
    | { type: "Int"; value: bigint }
    | { type: "String"; value: string }
    | { type: "Bytes"; value: Uint8Array }
    | { type: "Ident"; value: string }
    | { type: "LBrace" }
    | { type: "RBrace" }
    | { type: "LBracket" }
    | { type: "RBracket" }
    | { type: "Colon" }
    | { type: "Comma" }
    | { type: "EOF" };

const I64_MIN = -(1n << 63n);
const I64_MAX = (1n << 63n) - 1n;

export class Lexer {
    private offset = 0;
    private line = 1;
    private column = 1;

    constructor(private readonly input: string) { }

    private span(): Span {
        return { offset: this.offset, line: this.line, column: this.column };
    }

    private err(kind: any): ParseError {
        const span = this.span();
        return new ParseError(kind, span.offset, span.line, span.column);
    }

    private peek(): string | null {
        return this.offset < this.input.length ? this.input[this.offset]! : null;
    }

    private bump(): string | null {
        const char = this.peek();
        if (char === null) return null;

        this.offset++;

        if (char === "\n") {
            this.line++;
            this.column = 1;
        } else {
            this.column++;
        }

        return char;
    }

    private skipIgnored(): void {
        while (true) {
            // whitespace
            while (true) {
                const char = this.peek();
                if (char === " " || char === "\t" || char === "\n" || char === "\r") {
                    this.bump();
                } else {
                    break;
                }
            }

            // comment //
            if (this.peek() === "/" && this.input[this.offset + 1] === "/") {
                this.bump();
                this.bump();
                while (true) {
                    const char = this.peek();
                    if (char === null) {
                        break;
                    }
                    this.bump();
                    if (char === "\n") {
                        break;
                    }
                }
                continue;
            }

            // comment #
            if (this.peek() === "#") {
                this.bump();
                while (true) {
                    const char = this.peek();
                    if (char === null) {
                        break;
                    }
                    this.bump();
                    if (char === "\n") {
                        break;
                    }
                }
                continue;
            }

            break;
        }
    }

    private isIdentStart(char: string): boolean {
        return (
            (char >= "a" && char <= "z") ||
            (char >= "A" && char <= "Z") ||
            char === "_"
        );
    }

    private isIdentContinue(char: string): boolean {
        return this.isIdentStart(char) || (char >= "0" && char <= "9");
    }

    private lexIdentifier(): TokenKind {
        const start = this.offset;
        this.bump();

        while (true) {
            const char = this.peek();
            if (char !== null && this.isIdentContinue(char)) {
                this.bump();
            } else {
                break;
            }
        }

        const ident = this.input.slice(start, this.offset);

        if (ident === "null") return { type: "Null" };
        if (ident === "true") return { type: "True" };
        if (ident === "false") return { type: "False" };
        return { type: "Ident", value: ident };
    }

    private lexInt(): TokenKind {
        const start = this.offset;

        if (this.peek() === "-") this.bump();

        let sawDigit = false;
        while (true) {
            const char = this.peek();
            if (char !== null && char >= "0" && char <= "9") {
                sawDigit = true;
                this.bump();
            } else {
                break;
            }
        }

        if (!sawDigit) {
            throw this.err({ type: "IntegerOutOfRange" });
        }

        const text = this.input.slice(start, this.offset);

        let n: bigint;

        try {
            n = BigInt(text);
        } catch {
            throw this.err({ type: "IntegerOutOfRange" });
        }

        if (n < I64_MIN || n > I64_MAX) {
            throw this.err({ type: "IntegerOutOfRange" });
        }

        return { type: "Int", value: n };
    }

    private hexDigit(char: string): number | null {
        if (char >= "0" && char <= "9")
            return char.charCodeAt(0) - "0".charCodeAt(0);
        if (char >= "a" && char <= "f")
            return char.charCodeAt(0) - "a".charCodeAt(0) + 10;
        if (char >= "A" && char <= "F")
            return char.charCodeAt(0) - "A".charCodeAt(0) + 10;
        return null;
    }

    private lexBytes(): TokenKind {
        // must start with 0x
        if (!(this.peek() === "0" && this.input[this.offset + 1] === "x")) {
            throw this.err({ type: "MalformedBytesLiteral" });
        }

        this.bump(); // 0
        this.bump(); // x

        const hexStart = this.offset;

        while (true) {
            const char = this.peek();
            if (char !== null && this.hexDigit(char) !== null) {
                this.bump();
            } else {
                break;
            }
        }

        const hex = this.input.slice(hexStart, this.offset);

        if (hex.length === 0 || hex.length % 2 !== 0) {
            throw this.err({ type: "MalformedBytesLiteral" });
        }

        const out = new Uint8Array(hex.length / 2);
        for (let i = 0; i < hex.length; i += 2) {
            const hi = this.hexDigit(hex[i]!);
            const lo = this.hexDigit(hex[i + 1]!);
            if (hi === null || lo === null) {
                throw this.err({ type: "MalformedBytesLiteral" });
            }
            out[i / 2] = (hi << 4) | lo;
        }

        return { type: "Bytes", value: out };
    }

    private lexString(): TokenKind {
        this.bump(); // opening "

        let out = "";

        while (true) {
            const char = this.peek();
            if (char === null) {
                throw this.err({ type: "MalformedBytesLiteral" });
            }

            if (char === '"') {
                this.bump();
                return { type: "String", value: out };
            }

            if (char === "\\") {
                this.bump();
                const escape = this.bump();
                if (escape === null) {
                    throw this.err({ type: "MalformedStringLiteral" });
                }

                switch (escape) {
                    case '"':
                        out += '"';
                        break;
                    case "\\":
                        out += "\\";
                        break;
                    case "n":
                        out += "\n";
                        break;
                    case "r":
                        out += "\r";
                        break;
                    case "t":
                        out += "\t";
                        break;
                    case "u": {
                        // \uXXXX
                        let code = 0;
                        for (let i = 0; i < 4; i++) {
                            const hexChar = this.bump();
                            if (hexChar === null) {
                                throw this.err({ type: "MalformedStringLiteral" });
                            }
                            const digit = this.hexDigit(hexChar);
                            if (digit === null) {
                                throw this.err({ type: "MalformedStringLiteral" });
                            }
                            code = (code << 4) | digit;
                        }
                        out += String.fromCharCode(code);
                        break;
                    }
                    default:
                        throw this.err({ type: "MalformedStringLiteral" });
                }

                continue;
            }

            // Rust lexer rejects non-ASCII bytes in strings
            if (char.charCodeAt(0) >= 0x80) {
                throw this.err({ type: "MalformedStringLiteral" });
            }

            out += char;
            this.bump();
        }
    }

    nextToken(): Token {
        this.skipIgnored();
        const start = this.span();

        const char = this.peek();
        if (char === null) {
            return { kind: { type: "EOF" }, span: start };
        }

        const kind: TokenKind = (() => {
            switch (char) {
                case "{":
                    this.bump();
                    return { type: "LBrace" };
                case "}":
                    this.bump();
                    return { type: "RBrace" };
                case "[":
                    this.bump();
                    return { type: "LBracket" };
                case "]":
                    this.bump();
                    return { type: "RBracket" };
                case ":":
                    this.bump();
                    return { type: "Colon" };
                case ",":
                    this.bump();
                    return { type: "Comma" };
                case '"':
                    return this.lexString();
                case "0":
                    if (this.input[this.offset + 1] === "x") {
                        return this.lexBytes();
                    }
                    return this.lexInt();
                case "-":
                    return this.lexInt();
                default:
                    if (this.isIdentStart(char)) return this.lexIdentifier();
                    if (char >= "1" && char <= "9") return this.lexInt();
                    throw this.err({
                        type: "UnexpectedToken",
                        expected: "valid token",
                        found: "invalid character",
                    });
            }
        })();

        return { kind, span: start };
    }
}
